src/VisualStudio/Core/Def/Implementation/ProjectSystem/FileChangeWatcher.cs

```diff
@@ -149,6 +141,206 @@ public interface IFileWatchingToken
+                _directory = null!;
+                _filter = null;
+                _fileChangeFlags = 0;
+                _sink = null!;
+                _token = null!;
+                _tokens = null!;
+            }
+
+            private WatcherOperation(Kind kind, IEnumerable<IFileWatchingToken> tokens)
+            {
+                Contract.ThrowIfFalse(kind is Kind.UnwatchFiles);
+                _kind = kind;
+
+                _tokens = tokens;
+
+                // Other watching fields are not used for this kind
+                _directory = null!;
+                _filter = null;
+                _fileChangeFlags = 0;
+                _sink = null!;
+                _cookies = null!;
+                _token = null!;
+            }
+
+            private WatcherOperation(Kind kind, IFileWatchingToken token)
+            {
+                Contract.ThrowIfFalse(kind is Kind.UnwatchFile);
+                _kind = kind;
+
+                _token = token;
+
+                // Other watching fields are not used for this kind
+                _directory = null!;
+                _filter = null;
+                _fileChangeFlags = 0;
+                _sink = null!;
+                _cookies = null!;
+                _tokens = null!;
+            }
+
+            private enum Kind
+            {
+                None,
+                WatchDirectory,
+                WatchFile,
+                UnwatchFile,
+                UnwatchDirectories,
+                UnwatchFiles,
+            }
+
+            public static WatcherOperation Empty => new(Kind.None);
+
+            public static WatcherOperation WatchDirectory(string directory, string? filter, IVsFreeThreadedFileChangeEvents2 sink, List<uint> cookies)
+                => new(Kind.WatchDirectory, directory, filter, sink, cookies);
+
+            public static WatcherOperation WatchFile(string path, _VSFILECHANGEFLAGS fileChangeFlags, IVsFreeThreadedFileChangeEvents2 sink, IFileWatchingToken token)
+                => new(Kind.WatchFile, path, fileChangeFlags, sink, token);
+
+            public static WatcherOperation UnwatchDirectories(List<uint> cookies)
+                => new(Kind.UnwatchDirectories, cookies);
+
+            public static WatcherOperation UnwatchFiles(IEnumerable<IFileWatchingToken> tokens)
+                => new(Kind.UnwatchFiles, tokens);
+
+            public static WatcherOperation UnwatchFile(IFileWatchingToken token)
+                => new(Kind.UnwatchFile, token);
+
+            public bool TryCombineWith(in WatcherOperation other, out WatcherOperation combined)
+            {
+                if (other._kind == Kind.None)
+                {
+                    combined = this;
+                    return true;
+                }
+                else if (_kind == Kind.None)
+                {
+                    combined = other;
+                    return true;
+                }
```

Comment:
so combining really only helps when one is the none kind?  does that kind appear enough to matter?  if so, coolio.  if not, this doesn't seem that helpful.

Also... instead of this TryCombineWithPattern, couldn't we literally either ignore the .None operations, and make the batch loop much simpler?  it seems like a lot of pattern complexity to support combining if this is the end effective benefit.

Reply:
> so combining really only helps when one is the none kind?

I haven't implemented merging (that's the next change).

Reply:
will that be in this PR, or a followup?  Thanks!  if you do intend to do merging (now, or immediately in a followup PR), then i'm ok with this pattern remaining.

Reply:
I'm planning to implement all the things listed in the top post before marking this ready for review.

Reply:
➡️ Everything is implemented except the change to use two new bulk processing methods from `IVsAsyncFileChangeEx`. These changes will be submitted to **main-vs-deps** after this merges.